pages/partials/studentGroupControls.ejs

```diff
@@ -28,10 +28,33 @@
     Failed to create the group. It is already taken. Please try another one.
   </div>
 <% } %>
+
 <%- include('../shared/groupWorkInitial.ejs', {permissions: permissions}); %>
 <% } %>
 <% } else {%>
   <%- include('../shared/groupWorkInfoContainer.ejs', {permissions: permissions, group_info: group_info}); %>
+  <% if (usingGroupRoles && !rolesAreBalanced) { %>
+    <div class="alert alert-danger my-2" role="alert">
+      A user has too many roles. Every student should be assigned to exactly <strong>one</strong> role with group size <strong><%= groupsize %></strong> 
```

Comment:
My question here relates to my other question: can a given student have multiple roles in the same group? If so, this message is misleading. It's also not clear to me why the group size is relevant to an error message about a user having too many roles.

Backing up: how is it possible to get into a state where a user has "too many roles" - wouldn't we verify that that's not the case before inserting assignments into the DB?

Reply:
Regarding the unclear error message: we asked the SIIP team about their thoughts on the potential for group role configurations like the following, where one user has all the "required" roles (a role with a minimum > 0) and the rest of the users have non-required roles. Based on their feedback, we decided to prevent role configurations like these by enforcing that every user should have one role if `# of users >= # of required roles`.

![image](https://user-images.githubusercontent.com/55109467/205214380-130fd5ef-04de-43e6-9048-4126d193942b.png)

Admittedly, we didn't specify this requirement for users at all in the frontend (other than the error itself), so if this requirement still makes sense to implement, we can detail it in `studentGroupControls`.

Regarding the possibility of having "too many roles" in the first place: we originally wanted to verify that role assignments were valid before inserting them into the database. We first implemented this in the `POST` endpoint for the `update_group_roles` action, where we would call the `group_roles_validate` sproc to check if the role assignments were valid before inserting them into the database - if they weren't valid, they would never be inserted.

However, we couldn't find a way to cleanly display errors from the sproc to users from that `POST` endpoint - if the new role assignments were valid, we could just redirect to the same page, but if we wanted to add error messages to `res.locals`, we had to duplicate code from the `GET` endpoint to populate `res.locals` with the necessary variables as well as the validation error messages. 

Instead, we decided to allow *any* role assignment configuration to be inserted into the database, but validate whenever users load the page (in the `GET` endpoint) and subsequently disable users from starting the assessment if their group's role assignments were invalid. By validating within the `GET` endpoint, we could avoid duplicating code while also sending error messages if roles were invalid. This does mean that there could be invalid role assignments in the database, which is why we validate them when the page is loaded so we can disable the assessment from starting.

Do you think we should change our approach back to preventing invalid role assignments in the first place?

Reply:
Yeah, I apologize for the lack of the ability to show useful errors etc. in response to a `POST`. We'll be moving to a stateful session model soon, which should make this possible. In the meantime, I think the current validate-on-GET approach makes sense.

A related question: can roles be re-assigned once the assessment has been started? If so, what happens if the new configuration is no longer valid?

Reply:
Roles currently can't be reassigned once an assessment has started and no one can view the table, but the configuration should stay valid upon students leaving and joining as long as the assessment was started with a valid configuration. The SIIP team advised Renzo and I to include mid-assessment role reassignment in our next PR in order to keep the size of this current PR manageable and to start the review process sooner.

Is mid-assessment role reassignment something you'd like to see in this PR?